### PR TITLE
PR #45 (add/autoloadConfigurations)

### PR DESCRIPTION
**Pull request by rgadon107**
## PR Summary

`autoload_configurations()` loads a custom meta_box configuration into memory and initializes the custom fields configuration. The function is located in the `module.php` file of the `meta-data` handler within the Central Hub plugin. The function is invoked from a `config-loader.php` file within a custom plugin that leverages Central Hub. 

The path to the `module.php` file (relative to the Git repo root directory) is: 

>'/wp-content/mu-plugins/central-hub/src/meta-data/module.php`.

This PR contains the unit and integration tests ( with WordPress ) for `autoload_configurations()`.

### CODE REVIEW

**Comment by rgadon107** [comment]
Note: I am unable to mock 'KnowTheCode\ConfigStore\loadConfigFromFilesystem' using Brain\Monkey's `expect()` method and pass the function's 2 arguments. It returns the error message: 

>1) spiralWebDb\centralHub\Tests\Unit\Metadata\Tests_AutoloadConfigurations::test_should_store_a_meta_box_config_and_initialize_the_custom_fields_when_given_a_path_to_file
Mockery\Exception\InvalidCountException: Method KnowTheCode_ConfigStore_loadConfigFromFilesystem('path_to_config', 'defaults') from Mockery_0 should be called exactly 1 times but called 0 times.

However, Brain\Monkey's `when()` method works fine within the test method. I could not figure out why the `expect()` method failed to work. So I used `when()` and `justReturn()` instead.

File: mu-plugins/central-hub/tests/phpunit/fixtures/meta-box-members-runtime-config.php [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+				'default'      => '',
+				'delete_state' => '',
+				'sanitize'     => 'sanitize_text_field',
+			],
+			'is_active' => [
+				'default'      => 0,
+				'delete_state' => 0,
+				'sanitize'     => 'int_val',
+			]
+		],
```
Comment (hellofromtonya) [inline mu-plugins/central-hub/tests/phpunit/fixtures/meta-box-members-runtime-config.php]: @rgadon107 Notice that I changed the custom fields' configuration parameters and added a new one. 

Why? I did this to ensure that the initialized custom fields' configuration that gets stored in config store is as expected. 

What do I mean? `autoloadConfigurations()` does multiple things:

1. It uses `loadConfigFromFilesystem()` to pull the config file from the filesystem and then load it into the Config Store.
2. It uses `init_custom_fields_configuration()` to merge the default custom fields' configuration parameters and structure with the original config to ensure that the final custom fields' config has all of the parameters it needs.

To test that this works, we want a fixture (original config) that only has the custom parameters and not the default ones. 

Does that make sense?

You might be wondering why we do this. Right? To save code. The config files in each plugin are only required to specify the custom parameters. What does that mean? To save code, you only need to specify what's different. For example, you may have a checkbox (such as `is_active`) where it's configuration is different than a text field.

Does that make sense?

File: mu-plugins/central-hub/tests/phpunit/integration/meta-data/autoloadConfigurations.php [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+					'delete_state' => 0,
+					'sanitize'     => 'int_val',
+				],
+			],
+			'view'          => '',
+		];
+
+		// Check that each custom fields' configuration has all of the required parameters.
+		foreach ( $expected_config['custom_fields'] as $meta_key => $custom_field_config ) {
+			$this->assertSame( $custom_field_config, $actual_config['custom_fields'][ $meta_key ] );
```
Comment (hellofromtonya) [inline mu-plugins/central-hub/tests/phpunit/integration/meta-data/autoloadConfigurations.php]: @rgadon107 Notice how I've constructed this assertion: I'm checking that each custom field has been populated with all of the required parameters. The fixture only has certain custom params. But after running through `autoload_configurations()` and more specifically through `init_custom_fields_configuration()`, the default params are merged with the custom ones. This assertion is checking that this merging happened as we expect.

File: mu-plugins/central-hub/tests/phpunit/unit/meta-data/autoloadConfigurations.php [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+	}
+
+	/**
+	 * Test autoload configurations() should load a metabox config from the filesystem into Config Store.
+	 */
+	public function test_should_load_config_from_filesystem_into_config_store() {
+		$config_file = CENTRAL_HUB_ROOT_DIR . '/tests/phpunit/fixtures/meta-box-members-runtime-config.php';
+
+		// Assert that loadConfigFromFilesystem() runs once and receives the expected parameters.
+		Monkey\Functions\expect( 'KnowTheCode\ConfigStore\loadConfigFromFilesystem' )
```
Comment (hellofromtonya) [inline mu-plugins/central-hub/tests/phpunit/unit/meta-data/autoloadConfigurations.php]: @rgadon107 Notice that this `expect` assertion is the test for this method. We're testing that the dependent function `loadConfigFromFilesystem()` runs only once and it receives the parameters we expect it to receive.

Does this make sense to you?

File: mu-plugins/central-hub/tests/phpunit/unit/meta-data/autoloadConfigurations.php [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+			->once()
+			->with( $config_file, $this->defaults )
+			->andReturn( 'meta_box.members' );
+		Monkey\Functions\when( 'spiralWebDB\Metadata\init_custom_fields_configuration' )->justReturn();
```
Comment (hellofromtonya) [inline mu-plugins/central-hub/tests/phpunit/unit/meta-data/autoloadConfigurations.php]: @rgadon107 When you use `when()`, it is not a test assertion. Rather, when it runs, it just returns `null` and prevents a fatal error.

File: mu-plugins/central-hub/tests/phpunit/unit/meta-data/autoloadConfigurations.php [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+		autoload_configurations( [ CENTRAL_HUB_ROOT_DIR . '/tests/phpunit/fixtures/meta-box-members-runtime-config.php' ] );
+
+		// This is a placeholder to avoid PHPUnit error. The "expect" above is the assertion in this test.
+		$this->assertTrue( true );
+	}
+
+	/**
+	 * Test autoload configurations() should load multiple config files from the filesystem into Config Store.
+	 */
+	public function test_should_load_multiple_config_files_from_filesystem_into_config_store() {
```
Comment (hellofromtonya) [inline mu-plugins/central-hub/tests/phpunit/unit/meta-data/autoloadConfigurations.php]: @rgadon107 The function can receive more than one file path to process. Right? This test makes sure it works as expected.

File: mu-plugins/central-hub/tests/phpunit/fixtures/meta-box-members-runtime-config.php (hunk 1 above)
Comment (rgadon107) [reply mu-plugins/central-hub/tests/phpunit/fixtures/meta-box-members-runtime-config.php]: @hellofromtonya Yes.

Comment (rgadon107) [reply mu-plugins/central-hub/tests/phpunit/fixtures/meta-box-members-runtime-config.php]: @hellofromtonya Yes.

Comment (rgadon107) [reply mu-plugins/central-hub/tests/phpunit/fixtures/meta-box-members-runtime-config.php]: @hellofromtonya Yes.

**Pull request merged**